Source/AliveLib/SnoozeParticle.cpp

```diff
@@ -106,6 +132,186 @@ SnoozeParticle* SnoozeParticle::vdtor_4B08D0(signed int flags)
+        int yInScreen = FP_GetExponent(field_2C_y - pCamPos->field_4_y);
+
+        for (int i = 0; i < ALIVE_COUNTOF(explosionVerts); i++)
+        {
+            Line_G2* pZExplosionLine = &field_A4_G2_lines[bufIdx][i];
+            LineG2_Init(pZExplosionLine);
+
+            int scaledLineRelativeStartX = FP_GetExponent(FP_FromInteger(explosionVerts[i][0].field_0_x) * field_38_scale);
+            int scaledLineRelativeStartY = FP_GetExponent(FP_FromInteger(explosionVerts[i][0].field_2_y) * field_38_scale);
+            int scaledLineRelativeEndX = FP_GetExponent(FP_FromInteger(explosionVerts[i][1].field_0_x) * field_38_scale);
+            int scaledLineRelativeEndY = FP_GetExponent(FP_FromInteger(explosionVerts[i][1].field_2_y) * field_38_scale);
+            SetXY0(pZExplosionLine,
+                (__int16) PsxToPCX(xInScreen + scaledLineRelativeStartX, 11),
+                (__int16) (yInScreen + scaledLineRelativeStartY)
+            );
+            SetXY1(pZExplosionLine,
+                (__int16) PsxToPCX(xInScreen + scaledLineRelativeEndX, 11),
+                (__int16) (yInScreen + scaledLineRelativeEndY)
+            );
+
+            SetRGB0(pZExplosionLine,
+                (BYTE) (field_42_r / 2),
+                (BYTE) (field_44_g / 2),
+                (BYTE) (field_46_b / 2)
+            );
+            SetRGB1(pZExplosionLine,
+                (BYTE) field_42_r,
+                (BYTE) field_44_g,
+                (BYTE) field_46_b
+            );
+
+            Poly_Set_SemiTrans_4F8A60(&pZExplosionLine->mBase.header, 1);
+            OrderingTable_Add_4F8AA0(&pOt[field_40_layer], &pZExplosionLine->mBase.header);
+        }
+        rectToInvalidate.x = (__int16) PsxToPCX(xInScreen - 8, 0);
+        rectToInvalidate.w = (__int16) PsxToPCX(xInScreen + 8, 0);
+        rectToInvalidate.y = (__int16) (yInScreen - 8);
+        rectToInvalidate.h = (__int16) (yInScreen + 8);
+    }
+    else
+    {
+        Line_G4* pZLine = &field_4C_G4_lines[bufIdx];
+        LineG4_Init(pZLine);
+
+        __int16 xInScreen = FP_GetExponent(field_28_x - FP_FromInteger(FP_GetExponent(pCamPos->field_0_x)));
+        __int16 yInScreen = FP_GetExponent(field_2C_y - FP_FromInteger(FP_GetExponent(pCamPos->field_4_y)));
+
+        __int16 RectX_v_Psx = xInScreen + FP_GetExponent(FP_FromInteger(zVerts[0]) * field_38_scale);
+        __int16 RectW_v_Psx = xInScreen + FP_GetExponent(FP_FromInteger(zVerts[5]) * field_38_scale);
+
+        __int16 rectX_v = PsxToPCX(RectX_v_Psx, 11);
+        __int16 rectY_v = FP_GetExponent(FP_FromInteger(zVerts[1]) * field_38_scale) + yInScreen;
+        __int16 rectW_v = PsxToPCX(RectW_v_Psx, 11);
+        __int16 rectH_v = yInScreen + FP_GetExponent(FP_FromInteger(zVerts[7]) * field_38_scale);
+
+        SetXY0(pZLine,
+            rectX_v,
+            rectY_v
+        );
+        SetXY1(pZLine,
+            (__int16) PsxToPCX(xInScreen + FP_GetExponent(FP_FromInteger(zVerts[2]) * field_38_scale), 11),

```

Comment:
static cast these, or change xInScreen to the correct type so a cast isn't required